clients/harvester/client.go

```diff
@@ -0,0 +1,347 @@
+	if err != nil {
+		return err
+	}
+
+	defer resp.Body.Close()
+
+	if resp.StatusCode >= 300 {
+		return httperror.NewAPIErrorLong(resp.StatusCode, resp.Status, url)
+	}
+
+	byteContent, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return err
+	}
+
+	if len(byteContent) > 0 {
+		err = json.Unmarshal(byteContent, output)
+		if err != nil {
+			return err
+		}
+		return nil
+	}
+	return fmt.Errorf("received empty response")
+}
+
+// AsUser accepts a user object, and then creates a token for said `user`. Then it instantiates and returns a Client using the token created.
+// This function uses the login action, and user must have a correct username and password combination.
+func (c *Client) AsUser(user *management.User) (*Client, error) {
+	returnedToken, err := c.login(user)
+	if err != nil {
+		return nil, err
+	}
+
+	return NewClient(returnedToken.Token, c.Session)
+}
+
+// ReLogin reinstantiates a Client to update its API schema. This function would be used for a non admin user that needs to be
+// "reloaded" inorder to have updated permissions for certain resources.
+func (c *Client) ReLogin() (*Client, error) {
+	return NewClient(c.restConfig.BearerToken, c.Session)
+}
+
+// WithSession accepts a session.Session and instantiates a new Client to reference this new session.Session. The main purpose is to use it
+// when created "sub sessions" when tracking resources created at a test case scope.
+func (c *Client) WithSession(session *session.Session) (*Client, error) {
+	return NewClient(c.restConfig.BearerToken, session)
+}
+
+// GetClusterCatalogClient is a function that takes a clusterID and instantiates a catalog client to directly communicate with that specific cluster.
+func (c *Client) GetClusterCatalogClient(clusterID string) (*catalog.Client, error) {
+	restConfig := *c.restConfig
+	restConfig.Host = fmt.Sprintf("https://%s/k8s/clusters/%s", c.restConfig.Host, clusterID)
+
+	catalogClient, err := catalog.NewForConfig(&restConfig, c.Session)
+	if err != nil {
+		return nil, err
+	}
+
+	return catalogClient, nil
+}
+
+// GetRancherDynamicClient is a helper function that instantiates a dynamic client to communicate with the harvester host.
+func (c *Client) GetRancherDynamicClient() (dynamic.Interface, error) {
+	dynamic, err := frameworkDynamic.NewForConfig(c.Session, c.restConfig)
+	if err != nil {
+		return nil, err
+	}
+	return dynamic, nil
+}
+
+// GetKubeAPIProvisioningClient is a function that instantiates a provisioning client that communicates with the Kube API of a cluster
+func (c *Client) GetKubeAPIProvisioningClient() (*kubeProvisioning.Client, error) {
+	provClient, err := kubeProvisioning.NewForConfig(c.restConfig, c.Session)
+	if err != nil {
+		return nil, err
+	}
+
+	return provClient, nil
+}
+
+// GetKubeAPIRKEClient is a function that instantiates a rke client that communicates with the Kube API of a cluster
+func (c *Client) GetKubeAPIRKEClient() (*kubeRKE.Client, error) {
```

Comment:
have a question on this as well, since the clusters will be created on rancher itself, do we need to have this function for harvester client? because we create clusters on rancher itself so we can make use of rancher client isn't it?